### ISSUE TITLE
Has anyone noticed Huginn's job queue backing up badly after an upgrade?

### ISSUE THREAD

**Issue by virtadpt** [huginn/huginn#2692]
I just recently upgraded my Huginn server; using rvm, Ruby is now running v2.6.3p62 and Huginn is running Git commit f3ff176ba39197d900687bb6349a48e0e2c95091.  Since doing so, I've been routinely seeing the job queue back up to the order of one or two thousand pending jobs at a time, several times a day.  Jobs are still processing and events are circulating inside of Huginn, it's just taking significantly longer now.

My system load isn't too bad, 1.45/1.48/1.57 with 2203 pending jobs.  Memory utilization in total is less than 50%.  I currently have 18 delayed job runners going at this time.  Network load isn't too bad, either, 88.99KB down/5.87KB up (which is much lower than usual).

To try to compensate I've dropped the value of DELAYED_JOB_MAX_RUNTIME back down to 2 minutes and increased the DELAYED_JOB_SLEEP_DELAY to 60 seconds.  SCHEDULER_FREQUENCY is set to 0.5 and the event expiration job is set to 4 hour cycles.

Has anyone seen this happen before?

**Comment by virtadpt** [comment]
Running `mytop` showed that It's MySQL running a bunch of DELETE queries to clear out a lot of agents' old events and stored log messages causing it.

**Comment by dsander** [comment]
I don't think we modified anything recently that would explain the behavior. Do you know which version you ran before the upgrade? To be honest I would expect `DELAYED_JOB_SLEEP_DELAY=60` making it worse because every worked would burst process a bunch of Agents every 60 seconds.

Is the deletion in mysql happening constantly?

**Comment by virtadpt** [comment]
I was running Huginn commit 51a0ca9ad3f8413939244451337398c9ac915e7f before the upgrade.  I upgraded to f3ff176ba39197d900687bb6349a48e0e2c95091.

Do you think I should re-set DELAYED_JOB_SLEEP_DELAY back to 20 or 30 seconds?

Yes, it is happening constantly.  I'm running mytop and 20/33 running threads (out of a maximum possible of 100 MySQL threads) are of the form `Query updating DELETE FROM `agent_logs` WHERE `agent_logs`.`agent_id` = ...`  My Huginn install is bouncing from between roughly 225 queued jobs to well over 2000 queued jobs (but it is processing them, just really slowly).

**Comment by dsander** [comment]
Hmm I could not find anything in the changes that would explain this. What is the full query that is running and how long does it take to finish, a couple of milliseconds or more like seconds per deletion?

You could try to set `AGENT_LOG_LENGTH` to a really high number (the default is 200) to disable the deletion.

**Comment by virtadpt** [comment]
Off the top of my head, I do not know (no shell access from $dayjob).  From running Mytop, I know that it was a DELETE, it was a per-agent query, and it was long enough that it scrolled off the edge of my (maximized) terminal.  When I get home tonight I can do some jiggery pokery to try to grab the full query when it runs.

I went through most of my agents and set the delete times back to 7 days (instead of tuning them by use case) and that seems to have fixed a lot of problems.  It's not like I'm hurting for disk space or RAM right now.  I also made a couple of tweaks to the MySQL configuration.

**Comment by dsander** [comment]
Mhm so maybe mysql was just overwhelmed with the deletion queries?

**Comment by virtadpt** [comment]
That seems to be the case.  Huginn sets up InnoDB tables, which are pretty I/O intensive when it comes to DELETEs because it has to update the write log, delete the row, and then update the indices on the table.  Spacing them out by a couple of days really helps; in fact, when I was fixing it up I re-nice(1)'d some of the MySQL threads so now it's running pretty snappily.  I plan on writing this up and posting it in the near future.
